base/sort.jl

```diff
@@ -1757,6 +1759,41 @@ julia> v
 """
 sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...)
 
+function sort(x::NTuple;
+              alg::Algorithm=defalg(x),
+              lt=isless,
+              by=identity,
+              rev::Union{Bool,Nothing}=nothing,
+              order::Ordering=Forward,
+              scratch::Union{Vector, Nothing}=nothing)
+    # Can't do this check with type parameters because of https://github.com/JuliaLang/julia/issues/56698
+    scratch === nothing || eltype(x) == eltype(scratch) || throw(ArgumentError("scratch has the wrong eltype"))
+    _sort(x, alg, ord(lt,by,rev,order), (;scratch))::typeof(x)
+end
```

Comment:
This is what I'm proposing. There's no unbound method parameter like this, `T` always has a defined value in the second method (and the first method doesn't use `T`). Although there's a bit of code duplication in the keyword argument defaults.

```suggestion
function sort(::Tuple{};
              alg::Algorithm=defalg(x),
              lt=isless,
              by=identity,
              rev::Union{Bool,Nothing}=nothing,
              order::Ordering=Forward,
              scratch::Union{Vector, Nothing}=nothing)
    ()
end
function sort(x::Tuple{T, Vararg{T}};
              alg::Algorithm=defalg(x),
              lt=isless,
              by=identity,
              rev::Union{Bool,Nothing}=nothing,
              order::Ordering=Forward,
              scratch::Union{Vector{T}, Nothing}=nothing) where {T}
    _sort(x, alg, ord(lt,by,rev,order), (;scratch))::typeof(x)
end
```

Reply:
Your suggestion will cause CI to fail. Please read the issues I liked above and/or review the CI failure in [9f2c4eb](https://github.com/JuliaLang/julia/pull/54494/commits/9f2c4eb1deb638bf488b19f274c6774ee66a1711). You can also make a separate PR to try it out yourself.

Reply:
I'm familiar with the linked issues - they're caused by unbound method static parameters, when each method is considered in isolation. That is not relevant to my suggestion, as there are no such method static parameters that are sometimes unbound in this suggestion.

Specifically, the first method has no static parameters, while the second method has only the single static parameter, `T`, whose value is always defined, for any possible invocation, even if the first method were deleted.

If you insist I'll make a separate PR to demonstrate.

Reply:
Possibly the dispatch I propose is overly strict regarding `scratch`, not sure if `scratch` with non-concrete element type is supposed to be allowed.

Reply:
> If you insist I'll make a separate PR to demonstrate.

Thanks. I would prefer you do it in a  separate PR.

Reply:
The [PR](https://github.com/JuliaLang/julia/pull/56731) that applies that diff is failing tests on all platforms. I'm not going to block this PR waiting to figure that out.